packages/react-instantsearch-dom/src/components/VoiceSearch.tsx

```diff
@@ -0,0 +1,147 @@
+import React, { Component } from 'react';
+import { translatable, Translate } from 'react-instantsearch-core';
+import { createClassNames } from '../core/utils';
+import voiceSearchHelper, {
+  VoiceSearchHelper,
+  VoiceListeningState,
+} from '../lib/voiceSearchHelper';
+const cx = createClassNames('VoiceSearch');
+
+type InnerComponentProps = {
+  status: string;
+  errorCode?: string;
+  isListening: boolean;
+  transcript?: string;
+  isSpeechFinal?: boolean;
+  isBrowserSupported: boolean;
+};
+
+type VoiceSearchProps = {
+  searchAsYouSpeak?: boolean;
+  refine: (query: string) => void;
+  translate: Translate;
+  buttonComponent?: React.FC<InnerComponentProps>;
+  statusComponent?: React.FC<InnerComponentProps>;
+};
+
+const getDefaultButtonInnerElement = (
+  status: string,
+  errorCode: string | undefined,
+  isListening: boolean
+) => {
+  if (status === 'error' && errorCode === 'not-allowed') {
+    return (
+      <>
+        <line x1="1" y1="1" x2="23" y2="23" />
+        <path d="M9 9v3a3 3 0 0 0 5.12 2.12M15 9.34V4a3 3 0 0 0-5.94-.6" />
+        <path d="M17 16.95A7 7 0 0 1 5 12v-2m14 0v2a7 7 0 0 1-.11 1.23" />
+        <line x1="12" y1="19" x2="12" y2="23" />
+        <line x1="8" y1="23" x2="16" y2="23" />
+      </>
+    );
+  }
+  return (
+    <>
+      <path
+        d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
+        fill={isListening ? 'currentColor' : ''}
+      />
+      <path d="M19 10v2a7 7 0 0 1-14 0v-2" />
+      <line x1="12" y1="19" x2="12" y2="23" />
+      <line x1="8" y1="23" x2="16" y2="23" />
+    </>
+  );
+};
+
+const DefaultButton: React.FC<InnerComponentProps> = ({
+  status,
+  errorCode,
+  isListening,
+}) => {
+  return (
+    <svg
+      xmlns="http://www.w3.org/2000/svg"
+      width="16"
+      height="16"
+      viewBox="0 0 24 24"
+      fill="none"
+      stroke="currentColor"
+      strokeWidth="2"
+      strokeLinecap="round"
+      strokeLinejoin="round"
+    >
+      {getDefaultButtonInnerElement(status, errorCode, isListening)}
+    </svg>
+  );
+};
+
+const DefaultStatus: React.FC<InnerComponentProps> = ({ transcript }) => (
+  <p>{transcript}</p>
+);
+
+class VoiceSearch extends Component<VoiceSearchProps, VoiceListeningState> {
+  private voiceSearch: VoiceSearchHelper;
+
+  constructor(props: VoiceSearchProps) {
+    super(props);
+    const { searchAsYouSpeak = false, refine } = props;
+    this.voiceSearch = voiceSearchHelper({
+      searchAsYouSpeak,
+      onQueryChange: query => refine(query),
+      onStateChange: () => {
+        this.setState(this.voiceSearch.getState());
```

Comment:
We never clean the listeners on unmount, which might throw an error e.g. the widget is currently listening but for some reasons, it's unmounted from the page. The widget still listens but once the event fires and trigger the listener React throws an error.

> Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.

Note that we might have the same issue with InstantSearch.js. With the Vanilla version is would probably not happen that often but with the Vue & Angular integration the chances are higher.

Reply:
Okay, trying to fix it [here](https://github.com/algolia/instantsearch.js/pull/3779) first.

Reply:
https://github.com/algolia/react-instantsearch/pull/2316/commits/8f8e971714ec61e24640311ee4b76a4ebdc6b5df

https://github.com/algolia/react-instantsearch/pull/2316/commits/92cf94f9bd44939dda23a94dc86a263de80e7e0b